// @robojs/moderation/.robo/build/events/interactionCreate/select-setup-logs.js
import { createSetupMessage } from '../../commands/mod/setup.js';
import { Selects } from '../../core/constants.js';
import { hasPermission, logAction } from '../../core/utils.js';
import { updateSettings } from '../../core/settings.js';
import { logger } from 'robo.js';
import { ChannelType, Colors } from 'discord.js';
export default (async (interaction)=>{
    // Only handle interaction meant for this file
    if (!interaction.isChannelSelectMenu() || interaction.customId !== Selects.ChannelLogs.id) {
        return;
    }
    await interaction.deferUpdate();
    // Validate permissions
    if (!hasPermission(interaction, 'ModerateMembers')) {
        logger.debug(`User @${interaction.user.username} does not have permission to change moderator logs channel`);
        return interaction.followUp({
            content: `You don't have permission to use this.`,
            ephemeral: true
        });
    }
    // Validate channel
    const channel = interaction.channels.first();
    if (!channel || channel.type !== ChannelType.GuildText) {
        return interaction.followUp({
            content: 'Please select a valid channel.',
            ephemeral: true
        });
    }
    // Add self to logs channel if needed
    const selfId = interaction.member?.user?.id;
    if (!channel.permissionsFor(selfId)?.has('ReadMessageHistory') || !channel.permissionsFor(selfId)?.has('SendMessages') || !channel.permissionsFor(selfId)?.has('ViewChannel')) {
        logger.debug(`Adding self to channel #${channel.name} for guild ${interaction.guildId}`);
        await channel.permissionOverwrites.edit(selfId, {
            ReadMessageHistory: true,
            SendMessages: true,
            ViewChannel: true
        });
    }
    // Set logs channel
    const newSettings = updateSettings(interaction.guildId, {
        logsChannelId: channel.id
    });
    // Update setup message
    logger.debug(`Set logs channel for guild ${interaction.guildId} to #${channel.name}`);
    const setupMessage = createSetupMessage(interaction, newSettings);
    await interaction.editReply({
        content: setupMessage.content,
        components: setupMessage.components
    });
    // Log action to modlogs channel
    logAction(interaction.guildId, {
        embeds: [
            {
                title: 'Channel set',
                description: `Moderation logs channel has been set to ${channel}`,
                color: Colors.Blurple,
                timestamp: new Date().toISOString(),
                footer: {
                    icon_url: interaction.user.displayAvatarURL(),
                    text: 'by @' + interaction.user.username
                }
            }
        ]
    });
});
