// @robojs/ai/.robo/build/engines/openai/assistant.js
import { _PREFIX } from "../../../../.robo/build/core/constants.js";
import { openai } from "./api.js";
import { Flashcore, color, logger } from "robo.js";
export class Assistant {
    constructor(data){
        this._data = data;
    }
    get data() {
        return this._data;
    }
    async thread(options) {
        const { threadId , threadMessage  } = options;
        // First let's load the thread data
        const openaiThreadId = await Flashcore.get(threadId, {
            namespace: _PREFIX + '/thread-index'
        });
        let thread = await Flashcore.get(openaiThreadId, {
            namespace: _PREFIX + '/thread'
        });
        const threadExists = !!thread // Do not re-assign this later!
        ;
        logger.debug(`Found cached thread for "${color.bold(threadId)}":`, thread);
        // No thread data? Try getting it from OpenAI and cache it
        if (!threadExists && openaiThreadId) {
            thread = await openai.getThread({
                thread_id: openaiThreadId
            });
            logger.debug(`Retrieved fresh thread data:`, thread);
        }
        // Create a new thread if it really doesn't exist
        let isThreadCreated = false;
        if (!thread) {
            thread = await openai.createThread({
                messages: [
                    threadMessage
                ],
                metadata: {
                    threadId: threadId
                }
            });
            isThreadCreated = true;
            logger.debug(`Created new thread:`, thread);
        }
        // Still no thread? That's not good, abort!
        if (!thread) {
            throw new Error(`Failed to create thread for "${threadId}"`);
        }
        // Index the thread and its OpenAI ID equivalent
        if (!threadExists) {
            logger.debug(`Caching thread data for ${threadId}:`, thread);
            await Flashcore.set(thread.id, thread, {
                namespace: _PREFIX + '/thread'
            });
            await Flashcore.set(threadId, thread.id, {
                namespace: _PREFIX + '/thread-index'
            });
        }
        return {
            thread,
            threadCreated: isThreadCreated
        };
    }
    toJSON() {
        return JSON.stringify(this._data, null, 2);
    }
    toString() {
        return this._data.id;
    }
}
