// @robojs/ai/.robo/build/commands/imagine.js
import { AI } from "../../../.robo/build/core/ai.js";
import { logger } from "../../../.robo/build/core/logger.js";
export const config = {
    description: 'Bring your creative visions to life!',
    options: [
        {
            name: 'prompt',
            description: 'Describe your imaginative scene or concept for artistic creation',
            required: true
        }
    ]
};
export default (async (interaction)=>{
    const prompt = interaction.options.get('prompt')?.value;
    const response = await AI.generateImage({
        prompt
    });
    const images = response?.images?.map((image)=>image.url);
    logger.debug(`Imagined ${images?.length} images:`, images);
    if (!images?.length) {
        logger.warn('No images were generated.');
        return 'No images were generated.';
    }
    // Fetch image url as buffer then return as file
    const buffers = await Promise.all(images.map(async (image)=>{
        const response = await fetch(image);
        return Buffer.from(await response.arrayBuffer());
    }));
    return {
        files: buffers
    };
});
